R/touchstone_demographic_dataset.R

```diff
@@ -0,0 +1,245 @@
+}
+
+extract_touchstone_demographic_dataset <- function(e, path, con) {
+
+  # Do we have any touchstone_demographic_dataset work to do?
+
+  if (!is.null(e$touchstone_demographic_dataset_csv)) {
+
+    e_tdd_csv <- e$touchstone_demographic_dataset_csv
+
+    # Test csv columns are correct.
+
+    if (length(names(e_tdd_csv)) != 3) {
+      stop("Incorrect column count in touchstone_demographic_dataset.csv")
+    }
+
+    if (any(sort(names(e_tdd_csv)) !=
+      c("demographic_source", "demographic_statistic_type", "touchstone"))) {
+        stop("Incorrect columns in touchstone_demographic_dataset.csv")
+    }
+
+    if (class(e_tdd_csv$demographic_source) != "character") {
+      stop(paste0("demographic_source in touchstone_demographic_dataset.csv ",
+                  "must be character"))
+    }
+
+    if (class(e_tdd_csv$demographic_statistic_type) != "character") {
+      stop(paste0("demographic_statistic_type in touchstone_demographic_",
+                  "dataset.csv must be character"))
+    }
+
+    # Lookup any existing touchstone_demographic_datasets, for all
+    # touchstones in the csv. (also attaches info about the demographic
+    # source, statistic type and dataset for our lookups)
+
+    e$db_tdd <- db_tdd_for_touchstones(con, e_tdd_csv$touchstone)
+
+    # Now lookup demographic_source, demographic_statistic_type and
+    # touchstone info for all given CSV rows that we might want to add.
+
+    e <- c(e, list(
+      db_dsrc = db_get(con, "demographic_source", "code",
+        unique(e_tdd_csv$demographic_source)),
+
+      db_dstype = db_get(con, "demographic_statistic_type", "code",
+        unique(e_tdd_csv$demographic_statistic_type)),
+
+      tdd_touchstones = db_get(con, "touchstone", "id",
+        unique(e_tdd_csv$touchstone))
+    ))
+
+  # Now the demographic_dataset (which is source, type). For the sake of
+  # simplicity, look for matches of either source, or type, rather than
+  # filtering to the exact combinations, which would be a bit complicated
+  # at this point, and not worth much. There aren't very many of these.
+
+    e <- c(e, list(db_dd = db_dd_info(con,
+      e_tdd_csv$demographic_statistic_type,
+      e_tdd_csv$demographic_source)))
+
+  # In the extract tests and the transform, I need to work with the demographic
+  # dataset, so will mash them together once here. Not exactly an extract, but
+  # a very simple thing that saves some repetition later.
+
+    # A mash to compare with existing touchstone_demographic_datasets
+
+    e$touchstone_demographic_dataset_csv$mash <- paste(
+      e_tdd_csv$touchstone, e_tdd_csv$demographic_source,
+      e_tdd_csv$demographic_statistic_type, sep = '\r')
+
+    e$db_tdd$mash <- paste(e$db_tdd$touchstone,
+                           e$db_tdd$dsource_code,
+                           e$db_tdd$dtype_code, sep = '\r')
+
+    # A mash to compare with existing demographic_datasets
+
+    e$touchstone_demographic_dataset_csv$ds_mash <- paste(
```

Comment:
do we not have a `mash` function we should be using here?

Reply:
My mash function in utils.R (at the moment) does all fields in a generic swoop; this one is a subset. I could parameterise the existing mash to take optionally do a subset I guess...

Reply:
Done - and discovered the first line of old mash function was doing nothing useful.
Also added a couple of tests for mash in test_utils.R